### PR TITLE
Counterexamples - fix integer parsing bug and improve code quality

### PR DESCRIPTION
**Pull request by Dargones**
This pull request does not add any new functionality to counterexamples but instead makes several code quality improvements and bug fixes. Here is a summary:

- Correctly handle unbounded integers, sequence indices, etc. (fixes https://github.com/dafny-lang/dafny/issues/2350)
- Use the native `Microsoft.Dafny.Type` in place of `DafnyModelType` to improve readability and reduce the amount of string processing.
- Escape special characters when outputting the counterexample and use UTF codes for characters outside the 32 to 126 range (this prevents some parsing issues and also disambiguates between similarly looking characters such as semicolon and greek question mark) 
- Fix edge cases in object field name and datatype destructor name identification
- Remove dead code left from the original counterexample tool

I also added relevant tests and made a change to test generation to support the move from DafnyModelType to Microsoft.Dafny.Type

<small>By submitting this pull request, I confirm that my contribution is made under the terms of the [MIT license](https://github.com/dafny-lang/dafny/blob/master/LICENSE.txt).</small>

### CODE REVIEW

File: Source/DafnyLanguageServer.Test/Various/CounterExampleTest.cs [hunk 1]
```diff
@@ -1003,61 +1003,74 @@ class Cla__ss {
...
+        datatype D = C(i:int) {{ 
+          predicate p() {{true}}
+        }}
+
+        method test(d: D) {{
+          if (d.p()) {{
+            assert d.i != 123;
+          }}
+        }}
+      }}".TrimStart();
```
Comment (MikaelMayer) [inline Source/DafnyLanguageServer.Test/Various/CounterExampleTest.cs:1039]:
```suggestion
      var source = @"
      module M {
        datatype D = C(i:int) {
          predicate p() {true}
        }

        method test(d: D) {
          if (d.p()) {
            assert d.i != 123;
          }
        }
      }".TrimStart();
```
Can you please remove the interpolant "$" on the other tests where it's not necessary? That will improve legibility.

File: Source/DafnyLanguageServer.Test/Various/CounterExampleTest.cs [hunk 2]
```diff
@@ -1003,61 +1003,74 @@ class Cla__ss {
...
+      var documentItem = CreateTestDocument(source);
+      await client.OpenDocumentAndWaitAsync(documentItem, CancellationToken);
+      var counterExamples = (await RequestCounterExamples(documentItem.Uri)).ToArray();
+      Assert.AreEqual(1, counterExamples.Length);
+      Assert.IsTrue(counterExamples[0].Variables.ContainsKey("c:M.C?"));
+      Assert.IsTrue(counterExamples[0].Variables["c:M.C?"] is
+        "(c1 := '\\u1023', c2 := '\\u1023')" or
+        "(c2 := '\\u1023', c1 := '\\u1023')");
     }
+
```
Comment (MikaelMayer) [inline Source/DafnyLanguageServer.Test/Various/CounterExampleTest.cs:1074]:
```suggestion
```

File: Source/DafnyServer/CounterExampleGeneration/DafnyModel.cs [hunk 3]
```diff
@@ -137,7 +139,7 @@ public class DafnyModel {
         } else if (fn.Name.StartsWith("#") && fn.Name.IndexOf('.') != -1 && fn.Name[1] != '#') {
           foreach (var tpl in fn.Apps) {
             var elt = tpl.Result;
-            datatypeValues.Add(elt, tpl);
+            datatypeValues[elt] = tpl;
```
Comment (MikaelMayer) [inline Source/DafnyServer/CounterExampleGeneration/DafnyModel.cs:142]: Before, you would be ensuring an exception if the element already existed. Can you please explain me why you are confident we don't need to check this?

File: Source/DafnyServer/CounterExampleGeneration/DafnyModelState.cs [hunk 4]
```diff
@@ -10,40 +10,42 @@
...
+    internal DafnyModelState(DafnyModel model, Model.CapturedState state) {
       Model = model;
       State = state;
       VarIndex = 0;
       vars = new();
       varMap = new();
-      varNamesMap = new();
-      skolems = new List<DafnyModelVariable>(SkolemVars());
+      varNameCount = new();
+      boundVars = new(BoundVars());
```
Comment (MikaelMayer) [inline Source/DafnyServer/CounterExampleGeneration/DafnyModelState.cs:43]: Thanks for this much more meaningful variable name.

File: Source/DafnyServer/CounterExampleGeneration/DafnyModelState.cs [hunk 5]
```diff
@@ -142,18 +145,16 @@ public class DafnyModelState {
         if (val == null) {
           continue; // This variable has no value in the model, so ignore it.
         }
+
         var vn = DafnyModelVariableFactory.Get(this, val, v, duplicate: true);
-        if (curVars.ContainsKey(v)) {
-          Model.RegisterLocalValue(vn.Name, val);
-        }
```
Comment (MikaelMayer) [inline Source/DafnyServer/CounterExampleGeneration/DafnyModelState.cs:148]: I don't know what this was doing before, but can you please explain me why you could delete it safely?

File: Source/DafnyServer/CounterExampleGeneration/DafnyModelVariable.cs [hunk 6]
```diff
@@ -6,6 +6,8 @@
 using System.Linq;
 using System.Text.RegularExpressions;
 using Microsoft.Boogie;
+using Microsoft.Dafny;
```
Comment (MikaelMayer) [inline Source/DafnyServer/CounterExampleGeneration/DafnyModelVariable.cs:9]: Can you please fix these nullable annotations warnings above that the CI raised?

**Review by MikaelMayer — CHANGES_REQUESTED** [review]
A few changes and questions, otherwise, I'm looking forward to enhance counter-examples!

File: Source/DafnyServer/CounterExampleGeneration/DafnyModelVariable.cs [hunk 7]
```diff
@@ -192,51 +200,58 @@ internal SeqVariable(DafnyModelState state, Model.Element element, string name,
...
       this.seqLength = seqLength;
     }
 
-    public void AddAtIndex(DafnyModelVariable e, int? index) {
+    public void AddAtIndex(DafnyModelVariable e, string? index) {
       if (index == null) {
         return;
       }
-      seqElements[(int)index] = e;
+      seqElements[index] = e;
```
[the remark this answers is not in the document]
Comment (Dargones) [reply Source/DafnyServer/CounterExampleGeneration/DafnyModelVariable.cs:234]: Done

File: Source/DafnyServer/CounterExampleGeneration/DafnyModelVariable.cs (hunk 6 above)
Comment (Dargones) [reply Source/DafnyServer/CounterExampleGeneration/DafnyModelVariable.cs:10]: Have added `#nullable enable` to fix this

File: Source/DafnyServer/CounterExampleGeneration/DafnyModelState.cs (hunk 5 above)
Comment (Dargones) [reply Source/DafnyServer/CounterExampleGeneration/DafnyModelState.cs:148]: This remained from the original version that existed before I refactored the counterexample extraction code. I think its purpose must have been some form of field name preprocessing but at this point it is dead code. The only thing that [`DafnyModel.RegisterLocalValue`](https://github.com/dafny-lang/dafny/blob/dc6e7f086a2fe4860d8885ad7a0ab80ac13a2b8e/Source/DafnyServer/CounterExampleGeneration/DafnyModel.cs#L824-L830) does is update the `localValue` dictionary, which is never used elsewhere, so I removed it. The two methods that `RegisterLocalValue` calls, which are `CompareFieldNames` and `GetNumber`, are static, non side-effecting and operate on strings, so I also remove these as part of this PR.

File: Source/DafnyServer/CounterExampleGeneration/DafnyModel.cs (hunk 3 above)
Comment (Dargones) [reply Source/DafnyServer/CounterExampleGeneration/DafnyModel.cs:143]: Type aliasing can lead to an element being associated with different datatype functions in the Model so we should not be throwing exceptions in this case. My goal is also to prevent exceptions from arising as much as possible - in some cases, for instance, the tool might not be able to figure out the type of a variable and return UnknownType but this will not break the execution and some counterexample will still be extracted, even if a portion of it will not have complete information. 

On related note, would VSCode extension crash if there were an exception in counterexample extraction? Provided this PR gets merged, I don't think there should be situations where it should throw and exception but adding a try-catch somewhere might still make sense to prevent issues in the future. Do you know what’s the best place for this would be.

**Comment by Dargones** [comment]
Thank you, @MikaelMayer! Have adopted the changes.

File: Source/DafnyServer/CounterExampleGeneration/DafnyModelState.cs (hunk 5 above)
Comment (MikaelMayer) [reply Source/DafnyServer/CounterExampleGeneration/DafnyModelState.cs:148]: Fair enough, thanks for the explanation.

File: Source/DafnyServer/CounterExampleGeneration/DafnyModel.cs (hunk 3 above)
Comment (MikaelMayer) [reply Source/DafnyServer/CounterExampleGeneration/DafnyModel.cs:143]: If the language server crashes because of an unhandled exception, many times the VSCode extension will not be responsive anymore.
So I understand why you want to avoid crashes and I'm with you on this refactoring.
I'm just wondering whether if you could store the information somewhere if for any reason the same key was assigned two different values. At least for example, in a log file, or at best in some log message about the counter-example sent via the counterexample API (so that it's easier to debug, and to handle later when the UI for counterexamples will change). I think the log file would be fine.

In general it's better to add the try-catch to the closest place where the exception might be thrown, if not, where it makes the most sense.

Comment (Dargones) [reply Source/DafnyServer/CounterExampleGeneration/DafnyModel.cs:142]: Adding a logging functionality would make sense! I would keep that separate from this PR, if possible, but I already have some logging functionality for test generation on my fork, so I might be adding it in the near future.

**Comment by Dargones** [comment]
Have now fixed all nullability warnings. Let me know if you would like me to make any other changes!

**Review by MikaelMayer — DISMISSED** [review]
Great job ! Thanks for your patience.
Also, can you please confirm this enhancement is not going to break clients?

**Review by MikaelMayer — APPROVED** [review]

**Comment by Dargones** [comment]
Thank you, @MikaelMayer! I don't add any new functionality, so it shouldn't break anything.
